analyses/doublet-detection/exploratory-notebooks/03_compare-benchmark-results.Rmd

```diff
@@ -0,0 +1,522 @@
+
+### Prepare data
+
+```{r}
+# modify existing consensus columns to only consider 2 methods
+doublet_df_list <- doublet_df_list |>
+  purrr::map(
+    \(dataset_df) {
+      
+      dataset_df <- dataset_df |>
+        dplyr::rowwise() |>
+        # Update column `consensus_call`
+        dplyr::mutate(
+          consensus_call = dplyr::if_else(
+            all(c(scdbl_prediction, cxds_prediction) == "doublet"),
+            "doublet", 
+            "singlet"
+          )
+        ) |>
+        # Add Update `consensus_class`
+        dplyr::mutate(
+          consensus_class = dplyr::case_when(
+            consensus_call == "doublet" ~ "doublet",
+            all(c(scdbl_prediction, cxds_prediction) == "singlet") ~ "singlet", 
+            .default = "ambiguous"
+          )
+         ) |>
+        # Add Update `confusion_call`
+        dplyr::mutate(
+          confusion_call = dplyr::case_when(
+            consensus_call == "doublet" && ground_truth == "doublet" ~ "tp",
+            consensus_call == "singlet" && ground_truth == "singlet" ~ "tn",
+            consensus_call == "doublet" && ground_truth == "singlet" ~ "fp",
+            consensus_call == "singlet" && ground_truth == "doublet" ~ "fn"
+          )
+        )
+      
+      return(dataset_df)
+    }
+  ) |> 
+  purrr::set_names(dataset_names)
+```
+
+
+### PCA
+
+This section plots the PCA for each dataset, clockwise from the top left:
+
+1. `scDblFinder` singlet/doublet calls
+2. `cxds` singlet/doublet calls
+3. Ground truth single/doublets
+4. Consensus _call_ droplets are either "doublet" or "singlet"
+5. Consensus _class_: droplets are either "doublet", "singlet", or "ambiguous"
+
+
+```{r, fig.width = 12, fig.height = 6}
+doublet_df_list |>
+  purrr::imap(
+    \(df, dataset) {
+      
+      pc_color_columns <- c("scDblFinder prediction" = "scdbl_prediction",
+                            "cxds prediction" = "cxds_prediction",
+                            "Ground truth" = "ground_truth", 
+                            "Consensus call" = "consensus_call")
+      
+      pc_plot_list <- pc_color_columns |> 
+        purrr::imap(
+          \(color_column, plot_title) {
+            plot_pca_calls(
+              df, 
+              color_column = !!sym(color_column), 
+              plot_type = "calls",
+              title = plot_title
+            ) + theme(legend.position = "none")}
+        ) 
+      # plot for consensus_class, which needs three colors
+      # note that the list name isn't actually used here
+      pc_plot_list$consensus_class <- plot_pca_calls(
+        df, 
+        color_column = consensus_class,
+        plot_type = "class",
+        title = "Consensus class"
+      )
+
+      patchwork::wrap_plots(pc_plot_list) +
+        plot_annotation(
+            glue::glue("PCA for {dataset}"), 
+            theme = theme(plot.title = element_text(size = 16))) + 
+        plot_layout(guides = "collect")
```

Comment:
I think same thing, just pull this out to a function at the top of the notebook. It can be really simple and just take the `pc_color_columns` and the `df` as input.

Reply:
Noting we discussed off github that I'm not going to port this into a function.